File: app/src/main/java/com/dewonderstruck/apps/ashx0/di/AppModule.java
package com.dewonderstruck.apps.ashx0.di;

import android.app.Application;
import android.content.SharedPreferences;
import androidx.preference.PreferenceManager;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.dewonderstruck.apps.Config;
import com.dewonderstruck.apps.ashx0.api.PSApiService;
import com.dewonderstruck.apps.ashx0.db.AboutUsDao;
import com.dewonderstruck.apps.ashx0.db.BasketDao;
import com.dewonderstruck.apps.ashx0.db.BlogDao;
import com.dewonderstruck.apps.ashx0.db.CategoryDao;
import com.dewonderstruck.apps.ashx0.db.CategoryMapDao;
import com.dewonderstruck.apps.ashx0.db.CityDao;
import com.dewonderstruck.apps.ashx0.db.CommentDao;
import com.dewonderstruck.apps.ashx0.db.CommentDetailDao;
import com.dewonderstruck.apps.ashx0.db.CountryDao;
import com.dewonderstruck.apps.ashx0.db.DeletedObjectDao;
import com.dewonderstruck.apps.ashx0.db.HistoryDao;
import com.dewonderstruck.apps.ashx0.db.ImageDao;
import com.dewonderstruck.apps.ashx0.db.NotificationDao;
import com.dewonderstruck.apps.ashx0.db.PSAppInfoDao;
import com.dewonderstruck.apps.ashx0.db.PSAppVersionDao;
import com.dewonderstruck.apps.ashx0.db.PSCoreDb;
import com.dewonderstruck.apps.ashx0.db.ProductAttributeDetailDao;
import com.dewonderstruck.apps.ashx0.db.ProductAttributeHeaderDao;
import com.dewonderstruck.apps.ashx0.db.ProductCollectionDao;
import com.dewonderstruck.apps.ashx0.db.ProductColorDao;
import com.dewonderstruck.apps.ashx0.db.ProductDao;
import com.dewonderstruck.apps.ashx0.db.ProductMapDao;
import com.dewonderstruck.apps.ashx0.db.ProductSpecsDao;
import com.dewonderstruck.apps.ashx0.db.RatingDao;
import com.dewonderstruck.apps.ashx0.db.ShippingMethodDao;
import com.dewonderstruck.apps.ashx0.db.ShopDao;
import com.dewonderstruck.apps.ashx0.db.SubCategoryDao;
import com.dewonderstruck.apps.ashx0.db.TransactionDao;
import com.dewonderstruck.apps.ashx0.db.TransactionOrderDao;
import com.dewonderstruck.apps.ashx0.db.UserDao;
import com.dewonderstruck.apps.ashx0.utils.AppLanguage;
import com.dewonderstruck.apps.ashx0.utils.Connectivity;
import com.dewonderstruck.apps.ashx0.utils.LiveDataCallAdapterFactory;

import java.util.concurrent.TimeUnit;

import javax.inject.Singleton;

import androidx.room.Room;
import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;


/**
 * Created by Vamsi Madduluri on 11/15/17.
 * Contact Email : vamsii.wrkhost@gmail.com
 */

@Module(includes = ViewModelModule.class)
class AppModule {

    @Singleton
    @Provides
    PSApiService providePSApiService() {

        Gson gson = new GsonBuilder()
                .setDateFormat("yyyy-MM-dd HH:mm:ss")
                .create();

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .writeTimeout(1, TimeUnit.MINUTES)
                .readTimeout(1, TimeUnit.MINUTES)
                .build();

        return new Retrofit.Builder()
                .baseUrl(Config.APP_API_URL)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                .build()
                .create(PSApiService.class);

    }

    @Singleton
    @Provides
    PSCoreDb provideDb(Application app) {
        return Room.databaseBuilder(app, PSCoreDb.class, "PSApp.db")
                //.addMigrations(MIGRATION_1_2)
                .fallbackToDestructiveMigration()
                .build();
    }

    @Singleton
    @Provides
    Connectivity provideConnectivity(Application app) {
        return new Connectivity(app);
    }

    @Singleton
    @Provides
    SharedPreferences provideSharedPreferences(Application app) {
        return PreferenceManager.getDefaultSharedPreferences(app.getApplicationContext());
    }

    @Singleton
    @Provides
    UserDao provideUserDao(PSCoreDb db) {
        return db.userDao();
    }

    @Singleton
    @Provides
    AppLanguage provideCurrentLanguage(SharedPreferences sharedPreferences) {
        return new AppLanguage(sharedPreferences);
    }

    @Singleton
    @Provides
    AboutUsDao provideAboutUsDao(PSCoreDb db) {
        return db.aboutUsDao();
    }

    @Singleton
    @Provides
    ImageDao provideImageDao(PSCoreDb db) {
        return db.imageDao();
    }

    @Singleton
    @Provides
    ProductDao provideProductDao(PSCoreDb db) {
        return db.productDao();
    }

    @Singleton
    @Provides
    CountryDao provideCountryDao(PSCoreDb db) {
        return db.countryDao();
    }

    @Singleton
    @Provides
    CityDao provideCityDao(PSCoreDb db) {
        return db.cityDao();
    }

    @Singleton
    @Provides
    ProductColorDao provideProductColorDao(PSCoreDb db) {
        return db.productColorDao();
    }

    @Singleton
    @Provides
    ProductSpecsDao provideProductSpecsDao(PSCoreDb db) {
        return db.productSpecsDao();
    }

    @Singleton
    @Provides
    ProductAttributeHeaderDao provideProductAttributeHeaderDao(PSCoreDb db) {
        return db.productAttributeHeaderDao();
    }

    @Singleton
    @Provides
    ProductAttributeDetailDao provideProductAttributeDetailDao(PSCoreDb db) {
        return db.productAttributeDetailDao();
    }

    @Singleton
    @Provides
    BasketDao provideBasketDao(PSCoreDb db) {
        return db.basketDao();
    }

    @Singleton
    @Provides
    HistoryDao provideHistoryDao(PSCoreDb db) {
        return db.historyDao();
    }

    @Singleton
    @Provides
    CategoryDao provideCategoryDao(PSCoreDb db) {
        return db.categoryDao();
    }

    @Singleton
    @Provides
    RatingDao provideRatingDao(PSCoreDb db) {
        return db.ratingDao();
    }

    @Singleton
    @Provides
    SubCategoryDao provideSubCategoryDao(PSCoreDb db) {
        return db.subCategoryDao();
    }

    @Singleton
    @Provides
    CommentDao provideCommentDao(PSCoreDb db) {
        return db.commentDao();
    }

    @Singleton
    @Provides
    CommentDetailDao provideCommentDetailDao(PSCoreDb db) {
        return db.commentDetailDao();
    }

    @Singleton
    @Provides
    NotificationDao provideNotificationDao(PSCoreDb db){return db.notificationDao();}

    @Singleton
    @Provides
    ProductCollectionDao provideProductCollectionDao(PSCoreDb db){return db.productCollectionDao();}

    @Singleton
    @Provides
    TransactionDao provideTransactionDao(PSCoreDb db){return db.transactionDao();}

    @Singleton
    @Provides
    TransactionOrderDao provideTransactionOrderDao(PSCoreDb db){return db.transactionOrderDao();}

//    @Singleton
//    @Provides
//    TrendingCategoryDao provideTrendingCategoryDao(PSCoreDb db){return db.trendingCategoryDao();}

    @Singleton
    @Provides
    ShopDao provideShopDao(PSCoreDb db){return db.shopDao();}

    @Singleton
    @Provides
    BlogDao provideNewsFeedDao(PSCoreDb db){return db.blogDao();}

    @Singleton
    @Provides
    ShippingMethodDao provideShippingMethodDao(PSCoreDb db){return db.shippingMethodDao();}

    @Singleton
    @Provides
    ProductMapDao provideProductMapDao(PSCoreDb db){return db.productMapDao();}

    @Singleton
    @Provides
    CategoryMapDao provideCategoryMapDao(PSCoreDb db){return db.categoryMapDao();}

    @Singleton
    @Provides
    PSAppInfoDao providePSAppInfoDao(PSCoreDb db){return db.psAppInfoDao();}

    @Singleton
    @Provides
    PSAppVersionDao providePSAppVersionDao(PSCoreDb db){return db.psAppVersionDao();}

    @Singleton
    @Provides
    DeletedObjectDao provideDeletedObjectDao(PSCoreDb db){return db.deletedObjectDao();}
}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/di/AppModule.kt
package com.dewonderstruck.apps.ashx0.di

import android.app.Application
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import androidx.room.Room
import com.dewonderstruck.apps.Config
import com.dewonderstruck.apps.ashx0.api.PSApiService
import com.dewonderstruck.apps.ashx0.db.*
import com.dewonderstruck.apps.ashx0.utils.AppLanguage
import com.dewonderstruck.apps.ashx0.utils.Connectivity
import com.dewonderstruck.apps.ashx0.utils.LiveDataCallAdapterFactory
import com.google.gson.GsonBuilder
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit
import javax.inject.Singleton

/**
 * Created by Vamsi Madduluri on 11/15/17.
 * Contact Email : vamsii.wrkhost@gmail.com
 */
@Module(includes = [ViewModelModule::class])
internal class AppModule {
    @Singleton
    @Provides
    fun providePSApiService(): PSApiService {
        val gson = GsonBuilder()
                .setDateFormat("yyyy-MM-dd HH:mm:ss")
                .create()
        val okHttpClient = OkHttpClient.Builder()
                .writeTimeout(1, TimeUnit.MINUTES)
                .readTimeout(1, TimeUnit.MINUTES)
                .build()
        return Retrofit.Builder()
                .baseUrl(Config.APP_API_URL)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(LiveDataCallAdapterFactory())
                .build()
                .create(PSApiService::class.java)
    }

    @Singleton
    @Provides
    fun provideDb(app: Application?): PSCoreDb {
        return Room.databaseBuilder(app!!, PSCoreDb::class.java, "App.db") //.addMigrations(MIGRATION_1_2)
                .fallbackToDestructiveMigration()
                .build()
    }

    @Singleton
    @Provides
    fun provideConnectivity(app: Application?): Connectivity {
        return Connectivity(app)
    }

    @Singleton
    @Provides
    fun provideSharedPreferences(app: Application): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(app.applicationContext)
    }

    @Singleton
    @Provides
    fun provideUserDao(db: PSCoreDb): UserDao {
        return db.userDao()
    }

    @Singleton
    @Provides
    fun provideCurrentLanguage(sharedPreferences: SharedPreferences?): AppLanguage {
        return AppLanguage(sharedPreferences)
    }

    @Singleton
    @Provides
    fun provideAboutUsDao(db: PSCoreDb): AboutUsDao {
        return db.aboutUsDao()
    }

    @Singleton
    @Provides
    fun provideImageDao(db: PSCoreDb): ImageDao {
        return db.imageDao()
    }

    @Singleton
    @Provides
    fun provideProductDao(db: PSCoreDb): ProductDao {
        return db.productDao()
    }

    @Singleton
    @Provides
    fun provideCountryDao(db: PSCoreDb): CountryDao {
        return db.countryDao()
    }

    @Singleton
    @Provides
    fun provideCityDao(db: PSCoreDb): CityDao {
        return db.cityDao()
    }

    @Singleton
    @Provides
    fun provideProductColorDao(db: PSCoreDb): ProductColorDao {
        return db.productColorDao()
    }

    @Singleton
    @Provides
    fun provideProductSpecsDao(db: PSCoreDb): ProductSpecsDao {
        return db.productSpecsDao()
    }

    @Singleton
    @Provides
    fun provideProductAttributeHeaderDao(db: PSCoreDb): ProductAttributeHeaderDao {
        return db.productAttributeHeaderDao()
    }

    @Singleton
    @Provides
    fun provideProductAttributeDetailDao(db: PSCoreDb): ProductAttributeDetailDao {
        return db.productAttributeDetailDao()
    }

    @Singleton
    @Provides
    fun provideBasketDao(db: PSCoreDb): BasketDao {
        return db.basketDao()
    }

    @Singleton
    @Provides
    fun provideHistoryDao(db: PSCoreDb): HistoryDao {
        return db.historyDao()
    }

    @Singleton
    @Provides
    fun provideCategoryDao(db: PSCoreDb): CategoryDao {
        return db.categoryDao()
    }

    @Singleton
    @Provides
    fun provideRatingDao(db: PSCoreDb): RatingDao {
        return db.ratingDao()
    }

    @Singleton
    @Provides
    fun provideSubCategoryDao(db: PSCoreDb): SubCategoryDao {
        return db.subCategoryDao()
    }

    @Singleton
    @Provides
    fun provideCommentDao(db: PSCoreDb): CommentDao {
        return db.commentDao()
    }

    @Singleton
    @Provides
    fun provideCommentDetailDao(db: PSCoreDb): CommentDetailDao {
        return db.commentDetailDao()
    }

    @Singleton
    @Provides
    fun provideNotificationDao(db: PSCoreDb): NotificationDao {
        return db.notificationDao()
    }

    @Singleton
    @Provides
    fun provideProductCollectionDao(db: PSCoreDb): ProductCollectionDao {
        return db.productCollectionDao()
    }

    @Singleton
    @Provides
    fun provideTransactionDao(db: PSCoreDb): TransactionDao {
        return db.transactionDao()
    }

    @Singleton
    @Provides
    fun provideTransactionOrderDao(db: PSCoreDb): TransactionOrderDao {
        return db.transactionOrderDao()
    }

    //    @Singleton
    //    @Provides
    //    TrendingCategoryDao provideTrendingCategoryDao(PSCoreDb db){return db.trendingCategoryDao();}
    @Singleton
    @Provides
    fun provideShopDao(db: PSCoreDb): ShopDao {
        return db.shopDao()
    }

    @Singleton
    @Provides
    fun provideNewsFeedDao(db: PSCoreDb): BlogDao {
        return db.blogDao()
    }

    @Singleton
    @Provides
    fun provideShippingMethodDao(db: PSCoreDb): ShippingMethodDao {
        return db.shippingMethodDao()
    }

    @Singleton
    @Provides
    fun provideProductMapDao(db: PSCoreDb): ProductMapDao {
        return db.productMapDao()
    }

    @Singleton
    @Provides
    fun provideCategoryMapDao(db: PSCoreDb): CategoryMapDao {
        return db.categoryMapDao()
    }

    @Singleton
    @Provides
    fun providePSAppInfoDao(db: PSCoreDb): PSAppInfoDao {
        return db.psAppInfoDao()
    }

    @Singleton
    @Provides
    fun providePSAppVersionDao(db: PSCoreDb): PSAppVersionDao {
        return db.psAppVersionDao()
    }

    @Singleton
    @Provides
    fun provideDeletedObjectDao(db: PSCoreDb): DeletedObjectDao {
        return db.deletedObjectDao()
    }
}